frontend/linux/rpm/distro/dnf_install.go

```diff
@@ -0,0 +1,263 @@
+	return func(cfg *dnfInstallConfig) {
+		cfg.root = root
+	}
+}
+
+func DnfDownloadAllDeps(dest string) DnfInstallOpt {
+	return func(cfg *dnfInstallConfig) {
+		cfg.downloadOnly = true
+		cfg.allDeps = true
+		cfg.downloadDir = dest
+	}
+}
+
+func dnfInstallWithConstraints(opts []llb.ConstraintsOpt) DnfInstallOpt {
+	return func(cfg *dnfInstallConfig) {
+		cfg.constraints = opts
+	}
+}
+
+func dnfInstallFlags(cfg *dnfInstallConfig) string {
+	var cmdOpts string
+
+	if cfg.noGPGCheck {
+		cmdOpts += " --nogpgcheck"
+	}
+
+	if cfg.root != "" {
+		cmdOpts += " --installroot=" + cfg.root
+		cmdOpts += " --setopt=reposdir=/etc/yum.repos.d"
+	}
+
+	if cfg.downloadOnly {
+		cmdOpts += " --downloadonly"
+	}
+
+	if cfg.allDeps {
+		cmdOpts += " --alldeps"
+	}
+
+	if cfg.downloadDir != "" {
+		cmdOpts += " --downloaddir " + cfg.downloadDir
+	}
+
+	return cmdOpts
+}
+
+func dnfInstallOptions(cfg *dnfInstallConfig, opts []DnfInstallOpt) {
+	for _, o := range opts {
+		o(cfg)
+	}
+}
+
+func importGPGScript(keyPaths []string) string {
+	// all keys that are included should be mounted under this path
+	keyRoot := "/etc/pki/rpm-gpg"
+
+	var importScript string = "#!/usr/bin/env sh\nset -eux\n"
+	for _, keyPath := range keyPaths {
+		keyName := filepath.Base(keyPath)
+		importScript += fmt.Sprintf("gpg --import %s\n", filepath.Join(keyRoot, keyName))
+	}
+
+	return importScript
+}
+
+func TdnfInstall(cfg *dnfInstallConfig, releaseVer string, pkgs []string) llb.RunOption {
+	cmdFlags := dnfInstallFlags(cfg)
+	// tdnf makecache is needed to ensure that the package metadata is up to date if extra repo
+	// config files have been mounted
+	cmdArgs := fmt.Sprintf("set -ex; tdnf makecache; tdnf install -y --refresh --releasever=%s %s %s", releaseVer, cmdFlags, strings.Join(pkgs, " "))
+
+	var runOpts []llb.RunOption
+
+	// TODO(adamperlin): see if this can be removed for dnf
+	// If we have keys to import in order to access a repo, we need to create a script to use `gpg` to import them
+	// This is an unfortunate consequence of a bug in tdnf (see https://github.com/vmware/tdnf/issues/471).
+	// The keys *should* be imported automatically by tdnf as long as the repo config references them correctly and
+	// we mount the key files themselves under the right path. However, tdnf does NOT do this
+	// currently if the keys are referenced via a `file:///` type url,
+	// and we must manually import the keys as well.
+	if len(cfg.keys) > 0 {
+		importScript := importGPGScript(cfg.keys)
+		cmdArgs = "/tmp/import-keys.sh; " + cmdArgs
+		runOpts = append(runOpts, llb.AddMount("/tmp/import-keys.sh",
+			llb.Scratch().File(llb.Mkfile("/import-keys.sh", 0755, []byte(importScript))),
+			llb.SourcePath("/import-keys.sh")))
+	}
+
+	runOpts = append(runOpts, dalec.ShArgs(cmdArgs))
+	runOpts = append(runOpts, cfg.mounts...)
+
+	return dalec.WithRunOptions(runOpts...)
+
+}
+
+func DnfInstall(cfg *dnfInstallConfig, releaseVer string, pkgs []string) llb.RunOption {
```

Comment:
I've split out the install methods maybe prematurely, in case there are differences between dnf and tdnf. In particular, I'm not sure if the separate `/import-keys.sh` script is needed for dnf yet, since we haven't yet added distros that use dnf.